File: AF+SD+自带下拉刷新案例/YFTabBar/YFTabBar/Tab/YFTabbarController.m
//
//  YFTabbarController.m
//  YFTabbarController
//
//  Created by kangbing on 16/5/31.
//  Copyright © 2016年 kangbing. All rights reserved.
//

#import "YFTabbarController.h"
#import "OneViewController.h"
#import "TwoViewController.h"
#import "ThreeViewController.h"
#import "FourViewController.h"
#import "CenterViewController.h"
#import "YFNavigationController.h"
#import "YFTabbar.h"

@interface YFTabbarController ()

@end

@implementation YFTabbarController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    OneViewController *hvc = [[OneViewController alloc] init];
    [self addChildController:hvc title:@"首页" imageName:@"tab1-heartshow" selectedImageName:@"tab1-heart"];
    
    TwoViewController *fvc = [[TwoViewController alloc] init];
    [self addChildController:fvc title:@"活动" imageName:@"tab2-doctor" selectedImageName:@"tab2-doctorshow"];
    
    ThreeViewController *MoreVc = [[ThreeViewController alloc] init];
    [self addChildController:MoreVc title:@"更多" imageName:@"tab4-more" selectedImageName:@"tab4-moreshow"];
    
    FourViewController *svc = [[FourViewController alloc] init];
    [self addChildController:svc title:@"设置" imageName:@"tab5-file" selectedImageName:@"tab5-fileshow"];
    
    
    [[UITabBar appearance] setBackgroundImage:[self imageWithColor:[UIColor colorWithRed:248.0/255 green:248.0/255 blue:248.0/255 alpha:1.0]]];
    // 设置tabbar
    [[UITabBar appearance] setShadowImage:[[UIImage alloc] init]];
    // 设置自定义的tabbar
    [self setCustomtabbar];
    
    
    
}

- (void)setCustomtabbar{

    YFTabbar *tabbar = [[YFTabbar alloc]init];
    
    tabbar.vcCount = (int)self.viewControllers.count;
    
    [self setValue:tabbar forKeyPath:@"tabBar"];

    [tabbar.centerBtn addTarget:self action:@selector(centerBtnClick:) forControlEvents:UIControlEventTouchUpInside];


}

- (void)centerBtnClick:(UIButton *)button{
    
    CenterViewController *centerViewController = [[CenterViewController alloc]init];
    
    YFNavigationController *centerNavViewController = [[YFNavigationController alloc]initWithRootViewController:centerViewController];

    [self presentViewController:centerNavViewController animated:YES completion:nil];
    
}




- (void)addChildController:(UIViewController*)childController title:(NSString*)title imageName:(NSString*)imageName selectedImageName:(NSString*)selectedImageName
{
    
    childController.title = title;
    childController.tabBarItem.image = [[UIImage imageNamed:imageName] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    childController.tabBarItem.selectedImage = [[UIImage imageNamed:selectedImageName] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    // 设置一下选中tabbar文字颜色
    [childController.tabBarItem setTitleTextAttributes:@{ NSForegroundColorAttributeName : [UIColor darkGrayColor] } forState:UIControlStateSelected];
    
    YFNavigationController * nav = [[YFNavigationController alloc] initWithRootViewController:childController];
    
    [self addChildViewController:nav];
}


- (UIImage *)imageWithColor:(UIColor *)color{
    // 一个像素
    CGRect rect = CGRectMake(0, 0, 1, 1);
    // 开启上下文
    UIGraphicsBeginImageContextWithOptions(rect.size, NO, 0);
    [color setFill];
    UIRectFill(rect);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}


@end

File: 自定义TabBar 适配X  Swift/YFTabBar/YFTabBarController/Classes/Tab/YFTabBarController.swift
//
//  YFTabBarController.swift
//  YFTabBarController
//
//  Created by yangfan on 2018/3/15.
//  Copyright © 2018年 yangfan. All rights reserved.
//

import UIKit

class YFTabBarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        addChildViewController(childController: OneViewController(), title: "首页", imageName: "tab1-heartshow", selectedImageName: "tab1-heart")
        
        addChildViewController(childController: TwoViewController(), title: "活动", imageName: "tab2-doctor", selectedImageName: "tab2-doctorshow")
        
        addChildViewController(childController: ThreeViewController(), title: "设置", imageName: "tab4-more", selectedImageName: "tab4-moreshow")
        
        addChildViewController(childController: FourViewController(), title: "更多", imageName: "tab5-file", selectedImageName: "tab5-fileshow")
        
        let tabBar = YFTabBar()
        
        tabBar.vcCount = viewControllers?.count
        
        tabBar.centerBtn.addTarget(self, action: #selector(centerBtnClick), for: UIControlEvents.touchUpInside)
        
        UITabBar.appearance().backgroundImage = imageWithColor(color: UIColor(red: 248.0/255.0, green: 248.0/255.0, blue: 248.0/255.0, alpha: 1.0))

        UITabBar.appearance().shadowImage = UIImage()
        
        setValue(tabBar, forKey: "tabBar")
        
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}


extension YFTabBarController{
    
    func addChildViewController(childController: UIViewController, title:String, imageName:String, selectedImageName:String) {
        
        childController.title = title
        
        childController.tabBarItem.image = UIImage(named: imageName)?.withRenderingMode(UIImageRenderingMode.alwaysOriginal)
        
        childController.tabBarItem.selectedImage = UIImage(named: selectedImageName)?.withRenderingMode(UIImageRenderingMode.alwaysOriginal)
        
    childController.tabBarItem.setTitleTextAttributes([NSAttributedStringKey.foregroundColor : UIColor.darkGray], for: UIControlState.selected)
        
        
        addChildViewController(YFNavigationController(rootViewController: childController))
        
  
    }
    
    @objc func centerBtnClick(){
        
        present(CenterViewController(), animated: true, completion: nil)
        
    }
    
    func imageWithColor(color:UIColor) -> UIImage? {
        // 一个像素
        let rect = CGRect(x: 0, y: 0, width: 1, height: 1)
        // 开启上下文
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0)
        color.setFill()
        UIRectFill(rect);
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext();
        
        return image
    }
    
    
}
